### DISCUSSION TITLE
Deployment to PYNQ v2.7.0: execute_onnx(model, input_dict) fails

### DISCUSSION THREAD

**Discussion by luisbaehr** [Xilinx/finn#492] [Q&A]
Hi everyone,

I was trying to deploy the model of "tfc_end2end_example" to a PYNQ-Z1 board running PYNQ version 2.7.0, but the "execute_onnx()" function fails. I am running the example as the Jupyter notebook hosted by the FINN docker. When deploying to the board booted with a PYNQ v2.6.0 image it works without any problems.

I am pretty sure that I have set everything up, as explained in https://finn-dev.readthedocs.io/en/stable/getting_started.html
- IP for PYNQ is 192.168.2.99, IP for Host is 192.168.2.1
- bitstring package is installed through pip
- ssh key is installed

The error message of the Jupyter notebook is attached below.
![image](https://user-images.githubusercontent.com/97697477/149402073-14ca2383-543a-4797-ba9c-af767cc5710b.png)

As it seems to me that the execution on the PYNQ isn't happening at all, I tried to run the driver.py script directly on the PYNQ through ssh, but there are permissions missing when trying to run as "python3 driver.py" and the module "pynq" is missiong when run as "sudo python3 driver.py". Screenshots below.
![image](https://user-images.githubusercontent.com/97697477/149402796-e0bbe1b3-9632-49d9-848c-948e62db26d1.png)
![image](https://user-images.githubusercontent.com/97697477/149402868-5523b065-879f-47cd-9521-6cb5fb3b32cc.png)

Has anyone already encountered this problem? Is there a workaround, or should I just keep using PYNQ v2.6.0?

Best regards,
Luis

**Comment by rpitonak** [comment]
Hello,

Try to install `pynq` via `sudo pip3 install pynq` . The package is maybe not installed for the root user in PYNQ v.2.7.0. See [release notes](https://github.com/Xilinx/PYNQ/releases/tag/v2.7.0):
> Pynq and Jupyter now execute in a virtual environment (venv)

**Reply by luisbaehr** [reply]
Hi,
Thank you for your answer, I have overlooked this detail. However this only explains the error message I get when trying to run `sudo python3 driver.py`. The problem that the example fails at `execute_onnx()`  still remains.

The output for `driver.py` when executed with sudo and with deactivated venv (and installation of the missing packages) is now as follows:
![image](https://user-images.githubusercontent.com/97697477/149523687-bf7301a7-0ed7-4a10-b91e-fe12d0d5bf9f.png)

But I don't know if this has anything to do with the problem itself.

**Reply by Luc-Dierick** [reply]
Hi Luisbaehr, have you found the solution to this problem?

**Reply by luisbaehr** [reply]
Hi,
No, I have not found a solution. But I didn't invest that much time into getting it to work. For now I am using the PYNQ v2.6.0 image.

**Reply by Luc-Dierick** [reply]
Ah, the 2.6.0 image download link is not available anymore. Would I be able to get it from you?

**Reply by luisbaehr** [reply]
You can get all images from the PYNQ github. There you can chose the one you need, on your hardware.
[PYNQ Releases](https://github.com/Xilinx/PYNQ/releases )

**Reply by heborras** [reply]
Hi @luisbaehr,

I have seen the same issue on the Ultra96-V2 with the 2.7 image. And it appears to be an issue with the `PYTHONPATH` in the `sudo` environment. In particular the path `/usr/local/share/pynq-venv/lib/python3.8/site-packages` appears to be missing when running `sudo python3`, however the `PYTHONPATH` is correct when switching to the root user completely with the following:
```
sudo -i
python3
```
I have no idea why this happens and if anyone with more knowledge about how `sudo` works could chime in, that would be amazing.

To remedy this you can explicitly add this path before execution like this:
`sudo PYTHONPATH=/usr/local/share/pynq-venv/lib/python3.8/site-packages python3 driver.py`
This appears to work fine on the Ultra96. Could you maybe test it on your device as well?

**Reply by ThomasCourtois** [reply]
Hello i have the same issue and even with your PYTHONPATH, it doesn't work in `user` environnement. 
![image](https://user-images.githubusercontent.com/48708489/179999227-60c965c0-32be-438b-aaae-23f85faf2be3.png)

However without using `sudo `but with `sudo su` enviro it works! but only if i put `export XILINX_XRT=/usr` before and not starting with `sudo `
![image](https://user-images.githubusercontent.com/48708489/180000394-46815136-e935-4036-aaf6-140109b3abe4.png)
